### ISSUE TITLE
pyre-check doesn't detect naive date vs. tz aware date type difference

### ISSUE THREAD

**Issue by puruzio** [facebook/pyre-check#553]
Why can't Pyre find this type error?

```
from datetime import datetime
import pytz

utc = pytz.utc
a = utc.localize(datetime.now())

print(a)
#2021-12-11 15:55:35.187927+00:00

print( datetime.now())
#2021-12-11 15:55:35.188028

if a > datetime.now():  # type error here. TypeError: can't compare offset-naive and offset-aware datetimes
    print("done")
```

**Comment by stroxler** [comment]
The loose answer is that the type system is only aware of what we tell it.

In the builtin python datetime library, localized and non-localized dates are instances of the same class, distinguished only by state rather than by type. The pytz stubs explicitly state that the type returned by `localized` is `datetime`:
https://github.com/python/typeshed/blob/master/stubs/pytz/pytz/__init__.pyi#L31

This is all pyre knows about, so as far as pyre is concerned localized and nonlocalized datetimes are the same.


It would be possible to make a more type-safe datetime library where localized and non-localized dates are represented by different classes and the library functions carefully distinguish them, but a typechecker like pyre can't make that happen downstream, the library and function annotations would need to do this within the type system.

It's possible such a library lready exists (I've never heard of one, but there are a lot of python libraries).

**Comment by puruzio** [comment]
Thanks for your kind response!
